source/Account.cpp

```diff
@@ -150,322 +219,448 @@ void Account::PayExtra(int mortgage, int64_t amount)
+			mortgage.MissPayment();
+			receipt.paidInFull = false;
+		}
+		else
+		{
+			payment = mortgage.MakePayment();
+			credits -= payment;
+			receipt.creditsPaid += payment;
 		}
-		out << "and " << Format::CreditString(it->second) << " in " << it->first + ".";
-	}
-	else
-	{
-		if(salariesPaid)
-			out << Format::CreditString(salariesPaid) << " in crew salaries"
-				<< ((mortgagesPaid || finesPaid || debtPaid || maintenancePaid) ? " and " : ".");
-		if(maintenancePaid)
-			out << Format::CreditString(maintenancePaid) << "  in maintenance"
-				<< ((mortgagesPaid || finesPaid || debtPaid) ? " and " : ".");
-		if(mortgagesPaid)
-			out << Format::CreditString(mortgagesPaid) << " in mortgages"
-				<< ((finesPaid || debtPaid) ? " and " : ".");
-		if(finesPaid)
-			out << Format::CreditString(finesPaid) << " in fines"
-				<< (debtPaid ? " and " : ".");
-		if(debtPaid)
-			out << Format::CreditString(debtPaid) << " in debt.";
 	}
-	return out.str();
-}
-
 
-
-const map<string, int64_t> &Account::SalariesIncome() const
-{
-	return salariesIncome;
+	return receipt;
 }
 
 
 
-int64_t Account::SalariesIncomeTotal() const
+Receipt Account::PayDebts()
 {
-	return accumulate(
-		salariesIncome.begin(),
-		salariesIncome.end(),
-		0,
-		[](int64_t value, const std::map<string, int64_t>::value_type &salary)
+	// This function needs to preserve:
+	//    1. The number of credits paid towards ALL debts
+	//    2. Whether ANY fine was not paid in full
+	Receipt receipt;
+	if(mortgages.empty()) return receipt;
+
+	for(Mortgage &mortgage : mortgages)
+	{
+		if(!(mortgage.Type() == "Debt")) continue;
+		int64_t payment = mortgage.Payment();
+		if(payment > credits)
 		{
-			return value + salary.second;
+			mortgage.MissPayment();
+			receipt.paidInFull = false;
 		}
-	);
-}
-
-
+		else
+		{
+			payment = mortgage.MakePayment();
+			credits -= payment;
+			receipt.creditsPaid += payment;
+		}
+	}
 
-void Account::SetSalaryIncome(const string &name, int64_t amount)
-{
-	if(amount == 0)
-		salariesIncome.erase(name);
-	else
-		salariesIncome[name] = amount;
+	return receipt;
 }
 
 
 
-int64_t Account::CrewSalariesOwed() const
+const string Account::GenerateMissedPaymentLogs(std::vector<Receipt> *receipts) const
 {
-	return crewSalariesOwed;
-}
+	ostringstream log;
 
+	if(!receipts->at(0).paidInFull)
+		log << "You could not pay all your crew salaries. ";
 
+	if(!receipts->at(1).paidInFull)
+		log << "You could not pay all your maintenance costs. ";
 
-void Account::PaySalaries(int64_t amount)
-{
-	amount = min(min(amount, crewSalariesOwed), credits);
-	credits -= amount;
-	crewSalariesOwed -= amount;
-}
-
-
+	if(Mortgages().size() > 0)
+	{
+		if(!receipts->at(2).paidInFull || !receipts->at(3).paidInFull || !receipts->at(4).paidInFull)
+			log << "You missed a mortgage payment. ";
+	}
 
-int64_t Account::MaintenanceDue() const
-{
-	return maintenanceDue;
+	return log.str();
 }
 
 
 
-void Account::PayMaintenance(int64_t amount)
+void Account::UpdateMortgages()
```

Comment:
Could we make this part of `PayMortgages()`. It feels very appropriate to just remove it from the list directly after the last full payment is made than in a separate function.

Reply:
Leaving this as a generic function is preferable, as there are at least 3 types of mortgages (Mortgages, Fines, and Debt) stored together by the game engine now that #9286 has been merged.

Reply:
Removing the Mortgage after it has been fully paid should be about 2 lines of code, within the function where it is relevant, without anyone new to the code needing to do "mental switching" (as quyykk described it) to understand how they relate to each-other.

If we need future leverage with common functions for mortgages, then we can introduce them once we need them.

At this moment I feel that this 4 lines-function is just an example of the earlier remark that the code is split in functions that are too small.

Reply:
Can we split the difference, and I remove the call as you're requesting, but leave the utility function in so it's ready to go later down the line?

Reply:
No.

If we leave the utility function without using it, then it would just sit there like dead code. I also don't expect it to be used later down the line because of the "mental switching" as mentioned before, and if we ever do want to use it, then we could easily implment it using 1 line of code (using something like [erase_if](https://en.cppreference.com/w/cpp/container/vector/erase2).